From 8aec225c6323e1d060fa50868432a1d5a5a92390 Mon Sep 17 00:00:00 2001
From: Justin Cichra <1342149+jrcichra@users.noreply.github.com>
Date: Tue, 24 Sep 2024 00:19:07 -0400
Subject: [PATCH] feat(api): add pprof endpoint

This will help debug running go2rtc instances.

By default the pprof endpoint is off.
It can be enabled via the api.pprof setting.
---
 internal/api/api.go | 13 +++++++++++++
 1 file changed, 13 insertions(+)

diff --git a/internal/api/api.go b/internal/api/api.go
index 86817bd0..aa8bdc42 100644
--- a/internal/api/api.go
+++ b/internal/api/api.go
@@ -6,6 +6,7 @@ import (
 	"fmt"
 	"net"
 	"net/http"
+	"net/http/pprof"
 	"os"
 	"strconv"
 	"strings"
@@ -30,6 +31,7 @@ func Init() {
 			TLSCert    string `yaml:"tls_cert"`
 			TLSKey     string `yaml:"tls_key"`
 			UnixListen string `yaml:"unix_listen"`
+			Pprof      bool   `yaml:"pprof"`
 		} `yaml:"api"`
 	}
 
@@ -43,6 +45,9 @@ func Init() {
 		return
 	}
 
+	// overwrite default mux with new mux to avoid pprof auto registering
+	http.DefaultServeMux = http.NewServeMux()
+
 	basePath = cfg.Mod.BasePath
 	log = app.GetLogger("api")
 
@@ -54,6 +59,14 @@ func Init() {
 	HandleFunc("api/restart", restartHandler)
 	HandleFunc("api/log", logHandler)
 
+	if cfg.Mod.Pprof {
+		HandleFunc("/debug/pprof/", pprof.Index)
+		HandleFunc("/debug/pprof/cmdline", pprof.Cmdline)
+		HandleFunc("/debug/pprof/profile", pprof.Profile)
+		HandleFunc("/debug/pprof/symbol", pprof.Symbol)
+		HandleFunc("/debug/pprof/trace", pprof.Trace)
+	}
+
 	Handler = http.DefaultServeMux // 4th
 
 	if cfg.Mod.Origin == "*" {
